source/fe/mapping_q_cache.cc

```diff
@@ -70,18 +74,63 @@ MappingQCache<dim, spacedim>::preserves_vertex_locations() const
 
 
 
+template <int dim, int spacedim>
+void
+MappingQCache<dim, spacedim>::initialize(
+  const Mapping<dim, spacedim> &      mapping,
+  const Triangulation<dim, spacedim> &triangulation)
+{
+  // FE and FEValues in the case they are needed
+  FE_Nothing<dim, spacedim> fe;
+  Threads::ThreadLocalStorage<std::unique_ptr<FEValues<dim, spacedim>>>
+    fe_values_all;
+
+  this->initialize(
+    triangulation,
+    [&](const typename Triangulation<dim, spacedim>::cell_iterator &cell) {
+      const auto mapping_q_generic =
+        dynamic_cast<const MappingQGeneric<dim, spacedim> *>(&mapping);
+      if (mapping_q_generic != nullptr &&
+          this->get_degree() == mapping_q_generic->get_degree())
+        {
+          return mapping_q_generic->compute_mapping_support_points(cell);
+        }
+      else
+        {
+          // get FEValues (thread-safe); in the case that this thread has not
+          // created a an FEValues object yet, this helper-function also
+          // creates one with the right quadrature rule
```

Comment:
I remember that we had problems with `WorkStream` before regarding threading when only using only a single `ThreadLocalStorage<FEValues>` instance rather than `ThreadLocalStorage<std::list<FEValues>>`. In the general case, it could happen that the code executed by a thread could itself call into threaded code (here, it could possibly be something in the constructor of `FEValues`), and then TBB would decide to suspend the current thread until the result of the spawned workers would complete. Since then the current thread is idle, it might start the work on another cell and try to acquire an `FEValues` again. I think we are lucky enough here, in the sense that we have the `fe_values.get() == nullptr` check here, and this would simply be false when TBB starts to get crazy and mixes inner parallelism with outer one. It feels a bit brittle here, but I guess this is the best we can do. It would of course be even better if we had a function in `Mapping` that could be evaluated on a set of points without all that extreme overhead of `FEValues`, so we could slip the thread local storage altogether.

Reply:
> we could slip the thread local storage altogether.

Note that we have this kind of function in the mapping interface: https://github.com/dealii/dealii/blob/20d2280dadce1e217e9b3b9b72a46637f49d3e2e/include/deal.II/fe/mapping.h#L476-L480
So the question is if we trust that this unstructured evaluation routine is fast enough. Given the optimizations we did for `MappingQGeneric` previously, and since I can't see that we would be using a faster tensor product structure in this case, I would say yes.

Reply:
I am not opposed to the idea. But, I might need `FEValues` in PR #11882 for the general case. So we move the discussion there.